File: test_input/sublang.py
from pico_process import SubLanguageBase, is_identifier, Local, Scope
from collections import Counter

class MySubLanguage(SubLanguageBase):
    # called to parse the sub-language from a string
    # (strings consist of raw pico-8 chars ('\0' to '\xff') - not real unicode)
    def __init__(self, str, on_error, **_):
        # our trivial language consists of space-separated tokens in newline-separated statements
        self.stmts = [stmt.split() for stmt in str.splitlines()]
        # we can report parsing errors:
        #on_error("Example")

    # these are utility functions for our own use:

    def is_global(self, token):
        # is the token a global in our language? e.g. sin / rectfill / g_my_global
        return is_identifier(token)

    def is_member(self, token):
        # is the token a member in our language? e.g. .my_member / .x
        return token.startswith(".") and self.is_global(token[1:])
        
    # for --lint:

    # called to get globals defined within the sub-language's code
    def get_defined_globals(self, **_):
        for stmt in self.stmts:
            if len(stmt) > 1 and stmt[1] == "<-": # our lang's assignment token
                yield stmt[0]

    # called to lint the sub-language's code
    def lint(self, builtins, globals, on_error, **_):
        for stmt in self.stmts:
            for token in stmt:
                if self.is_global(token) and token not in builtins and token not in globals:
                    on_error("Identifier '%s' not found" % token)
        # could do custom lints too

    # for --minify:

    # called to get all characters that won't get removed or renamed by the minifier
    # (aka, all characters other than whitespace and identifiers)
    # this is optional and doesn't affect correctness, but can slightly improve compressed size
    def get_unminified_chars(self, **_):
        for stmt in self.stmts:
            for token in stmt:
                if not self.is_global(token) and not self.is_member(token):
                    yield from token

    # called to get all uses of globals in the language's code
    def get_global_usages(self, **_):
        usages = Counter()
        for stmt in self.stmts:
            for token in stmt:
                if self.is_global(token):
                    usages[token] += 1
        return usages
        
    # called to get all uses of members (table keys) in the language's code
    def get_member_usages(self, **_):
        usages = Counter()
        for stmt in self.stmts:
            for token in stmt:
                if self.is_member(token):
                    usages[token[1:]] += 1
        return usages

    # only needed if your language supports locals:
    # called to get all uses of locals in the language's code.
    # should return a Counter dict similar to above, except the keys are 
    # Local objects, and their scope (Scope objects) have 2 extra fields:
    #   used_globals - a set of all global names used in that scope
    #                  or in any of its child scopes
    #   used_locals - a set of all locals (Local objects) that are both:
    #                 a) delcared in that scope or its parent scopes
    #                 b) used in that scope or any of its child scopes
    def get_local_usages(self, **_):
        # fake test, just to see that the code is accepted
        # (may be nice to have real test for this?)
        fake_scope = Scope()
        fake_local = Local("test", fake_scope)
        fake_scope.used_globals = self.get_global_usages().keys()
        fake_scope.used_locals = {fake_local}
        return {fake_local: 1}

    # called to rename all uses of globals/members/locals
    def rename(self, globals, members, locals, **_):
        for stmt in self.stmts:
            for i, token in enumerate(stmt):
                if self.is_global(token) and token in globals:
                    stmt[i] = globals[token]
                elif self.is_member(token) and token[1:] in members:
                    stmt[i] = members[token[1:]]

    # called (after rename) to return a minified string
    def minify(self, **_):
        return "\n".join(" ".join(stmt) for stmt in self.stmts)

class SplitKeysSubLang(SubLanguageBase):
    # parses the string
    def __init__(self, str, **_):
        self.data = [item.split("=") for item in str.split(",")]

    # counts usage of keys
    # (returned keys are ignored if they're not identifiers)
    def get_member_usages(self, **_):
        return Counter(item[0] for item in self.data if len(item) > 1)

    # renames the keys
    def rename(self, members, **_):
        for item in self.data:
            if len(item) > 1:
                item[0] = members.get(item[0], item[0])

    # formats back to string
    def minify(self, **_):
        return ",".join("=".join(item) for item in self.data)

# this is called to get a sub-languge class by name
def sublanguage_main(lang, **_):
    if lang == "evally":
        return MySubLanguage
    elif lang == "splitkeys":
        return SplitKeysSubLang
    elif lang == "empty":
        return SubLanguageBase

File: test_input/sublang.lua
local pico_process = python.import("pico_process")
local SubLanguageBase = pico_process.SubLanguageBase
local module = {}

-- helper function like split, but ignores empties
function split_nonempty(str, ch)
    local res = split(str, ch, false)
    while (del (res, "")); -- (or copy to new table)
    return res
end

MySubLanguage = python.class(SubLanguageBase)

-- called to parse the sub-language from a string
-- (strings consist of raw pico-8 chars ('\0' to '\xff') - not real unicode)
function MySubLanguage:__init(str, opts)
    -- we may have received args that can be used to customize the language (not used here)
    self.args = opts.args
    -- our trivial language consists of space-separated tokens in newline-separated statements
    local lines = split_nonempty(str, "\n")
    self.stmts = {}
    for line in all(lines) do
        add(self.stmts, split_nonempty(line, " "))
    end
    -- we can report parsing errors:
    -- opts.on_error("Example")
end

-- these are utility functions for our own use:

function MySubLanguage:is_global(token)
    -- is the token a global in our language? e.g. sin / rectfill / g_my_global
    return pico_process.is_identifier(token)
end

function MySubLanguage:is_member(token)
    -- is the token a member in our language? e.g. .my_member / .x
    return token[1] == "." and self:is_global(sub(token, 2))
end
    
function MySubLanguage:is_assignment(stmt)
    return #stmt > 1 and stmt[2] == "<-" -- our lang's assignment token
end

-- for --lint:

-- called to get globals defined (aka assigned to) within the sub-language's code
function MySubLanguage:get_defined_globals()
    local globals = {}
    for stmt in all(self.stmts) do
        if self:is_assignment(stmt) then
            add(globals, stmt[1])
        end
    end
    return globals
end

-- called to get globals used (aka read from) within the sub-language's code
function MySubLanguage:get_used_globals()
    local globals = {}
    for stmt in all(self.stmts) do
        local start = 1
        if (self:is_assignment(stmt)) start = 3 -- ignore assigned to globals

        for i=start,#stmt do
            local token = stmt[i]
            if (self:is_global(token)) add(globals, token)
        end
    end
    return globals
end

-- called to lint the sub-language's code
function MySubLanguage:lint(opts)
    local builtins = python.table(opts.builtins)
    local globals = python.table(opts.globals)

    for stmt in all(self.stmts) do
        for token in all(stmt) do
            if self:is_global(token) and not builtins[token] and not globals[token] then
                opts.on_error("Identifier '" .. token .. "' not found")
            end
        end
    end
    -- could do custom lints too
end

-- for --minify:

-- called to get all characters that won't get removed or renamed by the minifier
-- (aka, all characters other than whitespace and identifiers)
-- this is optional and doesn't affect correctness, but can slightly improve compressed size
function MySubLanguage:get_unminified_chars()
    local chars = {}
    for stmt in all(self.stmts) do
        for token in all(stmt) do
            if not self:is_global(token) and not self:is_member(token) then
                for ch in all(token) do
                    add(chars, ch) -- if this overflows, could also use python.list directly...
                end
            end
        end
    end
    return python.list(chars)
end

-- called to get all uses of globals in the language's code
function MySubLanguage:get_global_usages()
    local usages = {}
    for stmt in all(self.stmts) do
        for token in all(stmt) do
            if self:is_global(token) then
                usages[token] = (usages[token] or 0) + 1
            end
        end
    end
    return python.dict(usages)
end
    
-- called to get all uses of members (table keys) in the language's code
function MySubLanguage:get_member_usages()
    local usages = {}
    for stmt in all(self.stmts) do
        for token in all(stmt) do
            if self:is_member(token) then
                local member = sub(token, 2)
                usages[member] = (usages[member] or 0) + 1
            end
        end
    end
    return python.dict(usages)
end

-- local usages is too much for picoscript

-- called to rename all uses of globals/members/locals
function MySubLanguage:rename(opts)
    local globals = python.table(opts.globals)
    local members = python.table(opts.members)

    for stmt in all(self.stmts) do
        for i, token in ipairs(stmt) do
            if self:is_global(token) and globals[token] then
                stmt[i] = globals[token]
            elseif self:is_member(token) and members[sub(token, 2)] then
                stmt[i] = members[sub(token, 2)]
            end
        end
    end
end

-- called (after rename) to return a minified string
function MySubLanguage:minify()
    local lines = {}
    for stmt in all(self.stmts) do
        add(lines, table.concat(stmt, " ")) -- can just use table.concat
    end
    return table.concat(lines, "\n")
end

-----------

SplitKeysSubLang = python.class(SubLanguageBase)

-- parses the string
function SplitKeysSubLang:__init(str)
    self.data = {}
    for item in all(split(str, ",", false)) do
        add(self.data, split(item, "=", false))
    end
end

-- counts usage of keys
-- (returned keys are ignored if they're not identifiers)
function SplitKeysSubLang:get_member_usages()
    local usages = {}
    for item in all(self.data) do
        if #item > 1 then
            local ident = item[1]
            usages[ident] = (usages[ident] or 0) + 1
        end
    end
    return python.dict(usages)
end

-- renames the keys
function SplitKeysSubLang:rename(opts)
    local members = python.table(opts.members)

    for item in all(self.data) do
        if #item > 1 then
            local member = members[item[1]]
            if (member) item[1] = member
        end
    end
end

-- formats back to string
function SplitKeysSubLang:minify(opts)
    local lines = {}
    for item in all(self.data) do
        add(lines, table.concat(item, "="))
    end
    return table.concat(lines, ",")
end

-----------

ArgsSubLang = python.class(SubLanguageBase)

function ArgsSubLang:__init(str, opts)
    if (opts.args != "arg1 arg2") opts.on_error("wrong arg!")
end

-----------

-- this is called to get a sub-languge class by name
function module.sublanguage_main(lang)
    if (lang == "evally") return MySubLanguage
    if (lang == "splitkeys") return SplitKeysSubLang
    if (lang == "empty") return SubLanguageBase
    if (lang == "args") return ArgsSubLang
end

return module